### ISSUE TITLE
Segmentation fault in Carbon

### ISSUE THREAD

**Issue by dennisprudlo** [nunomaduro/larastan#889]
- Larastan Version: 0.7.10
- `--level` used: 1-8

### Description
When running larastan the proccess stops with a segmentation fault in some files.

```shell
app/Traits/Services.php
app/Traits/TwoFactorAuthentication.php
app/Traits/UniqueSecret.php
app/Mail/Users/Registered.php
zsh: segmentation fault  ./vendor/bin/phpstan analyze --debug
```

When analyzing only one of the first three files no errors occur, but when analyzing the `Registered.php` the segmentation fault happens (`phpstan analyze --debug app/Mail/Users/Registered.php`).

### Laravel code where the issue was found
It's a regular Mailable class and the segfault happens when the analyzer reaches the assignment of the `verificationUrl` variable inside the constructor. When I comment out the 4 lines and run larastan the file is being analyzed without errors.
```php
class Registered extends Mailable implements ShouldQueue
{
    use Queueable, SerializesModels;

    public User $user;

    public string $verificationUrl;

    public function __construct(User $user)
    {
        $this->user             = $user;
        $this->verificationUrl  = url()->temporarySignedRoute('verify.create', now()->addMinutes(30), [
            'user'  => $this->user,
            'email' => $this->user->email
        ]);
    }

    // ...
}
```

After changing the paths for some tests to a more specific directory (`app/Models`) the same segfault error occurs in another file. In a model the following function causes the segfault:

```php
public function hasDetailedBookedAt() : bool {
    return $this->booked_at->hour !== 0
        || $this->booked_at->minute !== 0
        || $this->booked_at->second !== 0;
}
```
When I remove the return statement or comment out those three lines the analyzer continues on (and then fails at the User model later on).

**Comment by szepeviktor** [comment]
Thank you for your report.
Could you copy the error message here?

**Comment by dennisprudlo** [comment]
What's in the first code block is all I get. Just a `zsh: segmentation fault  ./vendor/bin/phpstan analyze --debug` and then the process is terminated.

**Comment by canvural** [comment]
Hi,,

Sorry to hear you are getting segfaults. They are always a trouble.

From the code examples you shared I think the common part is accessing to model properties.

Does a simple code like
```php
function foo(User $user): string
{
	return $user->email;
}
```

also causes segfault?

Other than that there is not much we can do to help. You need to debug yourself. Maybe start debugging with `gdb` and try to get a stacktrace when the segfault occurs.

You can also install XDebug, setup some breakpoints and run PHPStan with `./vendor/bin/phpstan analyze --debug --xdebug` to debug step by step.

If it's really about model property access, possible places to setup breakpoints are https://github.com/nunomaduro/larastan/blob/master/src/Properties/ModelPropertyExtension.php and https://github.com/nunomaduro/larastan/blob/master/src/Properties/SchemaAggregator.php

**Comment by dennisprudlo** [comment]
In the second example: The `booked_at` is a datetime column. When I remove the `$casts` property the analyzer works.
```php
protected $casts = [
    'booked_at' => 'datetime',
];
```

Well I get `Cannot access property $hour on string.` but thats right, because without casting we get a formatted string from the column.

https://github.com/nunomaduro/larastan/blob/2b6879ded50ff6254ebea4dd4dc8a3af67e223f5/src/Properties/ModelPropertyExtension.php#L241-L243 So in here the `$realType` should have at least `\Illuminate\Support\Carbon` and `\Carbon\Carbon` as the union type. But yeah will need to debug that later and find out why the error happens.

**Comment by szepeviktor** [comment]
Wow! A Carbon segfault.
What is your PHP version and OS?

**Comment by dennisprudlo** [comment]
I currently use PHP 8.0.0 on macOS Big Sur 11.0.1 (Intel-based)

**Comment by szepeviktor** [comment]
Very exotic!

**Comment by szepeviktor** [comment]
If you have access to a linux box (or having $0.1 in your packet for a cloud instance for 1 hour) you can try running Larastan on PHP 7.4 on Debian.

I think we cannot do much here.

**Comment by szepeviktor** [comment]
https://www.php.net/downloads.php#v8.0.8

Simply upgrading your PHP engine may help!

**Comment by dennisprudlo** [comment]
Wow! Indeed upgrading the patch version helped. No segfaults with php 8.0.8.Thanks :)

**Comment by szepeviktor** [comment]
I share your joy.

> upgrading the patch version helped

That is why I wear a hard hat and do not use new major version software.

**Comment by Maxoulak** [comment]
Hi,
I was encountering the same error on Laravel 8 and php 7.3.9.
I also tried with php 7.4.X, but the same segmentation fault occured (for me, it was caused by the Nesbot\Carbon library).
However, I can't upgrade to php 8 for now, so I tried on a virtual machine (Ubuntu 18.04), and magic happens : no seg fault.
This issue is specific to MacOS, so no patch on this library should solved the problem.
If you face the same error, hope you'll save time thanks to this comment.
